### DISCUSSION TITLE
Q: Built-in Management Extensions(with Remote API) for .NET runtime?

### DISCUSSION THREAD

**Discussion by gelihu** [dotnet/runtime#39775] [General]
I am wondering, if there is a JMX like capability for .NET runtime?

> The Java Management Extensions (JMX) API is a standard API for management and monitoring of resources such as applications, devices, services, and the Java virtual machine.

I am aware of dotnet-trace and counters, but I do not know how to expose them.

> The JMX API includes remote access, so a remote management program can interact with a running application for these purposes.

Something like this would be amazing.

**Comment by davidfowl** [comment]
Yes what do you want to expose?

**Reply by gelihu** [reply]
CPU, Memory, exception rate and custom event counters.

I just saw this [blog](https://devblogs.microsoft.com/dotnet/introducing-dotnet-monitor/), this is great and Prom exporter is also there 😀 

My only concern is, a shared volume and a sidecar are used for the Kubernetes deployment. So, for every deployment do I need to have a sidecar and a shared volume for every deployment? or Can I customize the monitor to use a single volume?  Resources might be a concern here(just a personal opinion).

This is very exciting, but it would be better if they are directly exported on a TCP/IP port of the main process without needing to start another process, like the way JVM does(I am sorry, I keep on using JVM as an example; it is really a powerful feature); then, I can have a single exporter for all of the deployments(again, a personal opinion 😄 )
